src/Storage/Record.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace Drupal\entity_hierarchy\Storage;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+
+/**
+ * Defines a factory for creating a nested set storage handler for hierarchies.
+ */
+class Record {
+
+  protected int $id;
+
+  protected ?int $revision_id = NULL;
+
+  protected ?int $target_id;
+
+  protected int $weight;
+
+  protected int $depth;
+
+  protected string $type;
+
+  protected ?array $children = NULL;
+
+  public static function create(string $type, int $id, int $weight, int $depth, int $target_id = NULL) {
```

Comment:
if this is the only way to create one, we should make the constructor private to avoid `new Record` creating an invalid object

Reply:
Not sure what you mean. There's two use cases, one is to create the record for 'self' when doing ancestors, because the query doesn't include itself. The other is to facilitate unit testing.